web/modules/custom/joinup_notification/joinup_notification.behat.inc

```diff
@@ -346,60 +367,164 @@ class JoinupNotificationSubContext extends DrupalSubContextBase implements Drupa
+  }
+
+  /**
+   * Clicks on a link matching a pattern given the email text and the recipient.
+   *
+   * Note that if multiple messages have been sent to the same user, all of them
+   * will be checked but only the first instance of the link matching the
+   * pattern will be clicked.
+   *
+   * @param string $pattern
+   *   The pattern that the link should match to.
+   * @param string $mail
+   *   The recipient's mail.
+   * @param string $no_mail_message
+   *   The exception message to show when no mail was found that matches the
+   *   given e-mail address.
+   * @param string $no_match_message
+   *   The exception message to show when the given pattern was not found in the
+   *   messages sent to the given e-mail address.
+   * @param string|null $template
+   *   Optional message template to filter the messages by.
+   *
+   * @throws \Exception
+   *   If no mail was found linked to the given email or if the pattern did not
+   *   find any matches.
+   */
+  protected function assertMailLinkMatchingPattern(string $pattern, string $mail, string $no_mail_message, string $no_match_message, ?string $template = NULL): void {
+    // @todo Uncomment this line when TagTrait is available in DrupalExtension.
+    // @see https://github.com/jhedstrom/drupalextension/pull/510
+    // $this->assertEmailTagPresent();
     $found = FALSE;
     foreach (\Drupal::state()->get('system.test_mail_collector', []) as $mail_sent) {
-      if ($mail_sent['key'] === 'mail_change_verification' && $mail_sent['to'] === $mail) {
-        $found = TRUE;
-        break;
+      // Optionally filter by mail template.
+      if (!empty($template) && $mail_sent['key'] !== $template) {
+        continue;
+      }
+
+      // Filter out messages that do not match the given e-mail address.
+      if ($mail_sent['to'] !== $mail) {
+        continue;
+      }
+
+      // This is a flag that a message was found to have been sent to the user.
+      $found = TRUE;
+      if (!preg_match($pattern, $mail_sent['plain'], $match)) {
+        continue;
       }
-    }
 
-    if (!$found) {
-      throw new \Exception("No mail change verification E-mail has been sent to $mail.");
+      $this->visitPath($match[0]);
+      return;
     }
 
-    $pattern = '#https?://[^/].*/user/mail-change/[^/].*/[^/].*/[^/].*/.*#';
-    if (!preg_match($pattern, $mail_sent['plain'], $match)) {
-      throw new \Exception("The mail change verification E-mail doesn't contain a valid verification link.");
+    if ($found) {
+      throw new \Exception($no_match_message);
     }
+    throw new \Exception($no_mail_message);
+  }
 
-    $this->visitPath($match[0]);
+  /**
+   * Checks if the current scenario or feature has the @email tag.
+   *
+   * Call this in steps that use the test mail collector so that the developer
+   * is alerted if this tag is not present.
+   */
+  protected function assertEmailTagPresent(): void {
```

Comment:
I am fine with having something like this as the same idea is already used for the `@api` tag.
However, I have used the absence of the `@email` tag for debugging purposes, e.g. I knew that the assertion would fail, but I wanted to test the output of the mail in mailcatcher. 
Maybe use a warning instead of an assertion?

Reply:
If it is for debugging, you can just temporarily comment it out or put a `return;` statement at the top?

The idea behind this is to stop the test as early as possible if the tag is missing, so that the dev doesn't waste time.